### ISSUE TITLE
Loss behaviour

### ISSUE THREAD

**Issue by ridasalam** [dbolya/yolact#53]
Hi, I am trying to overfit the network on a custom dataset and I notice that the training mask loss never goes below 1. Is there a specific reason why this loss does not decrease much? (I started with the pre-trained weights from im700).

Do you have some log files from your trainings on COCO? I just want to know what's the expected behaviour of the network. I have a batch size of 4 and all the losses fluctuate a lot between the iterations.

**Comment by dbolya** [comment]
We do a lot of data augmentation which is likely why you can't get it to overfit. Here are some of the final iterations for im700:

```
Toward the start:
[  0]     240 || B: 4.553 | C: 6.594 | M: 5.163 | S: 1.210 | T: 17.519 || ETA: 6 days, 20:00:21 || timer: 0.648
[  0]     250 || B: 4.536 | C: 6.595 | M: 5.161 | S: 1.208 | T: 17.500 || ETA: 6 days, 19:30:03 || timer: 0.716
[  0]     260 || B: 4.540 | C: 6.594 | M: 5.174 | S: 1.199 | T: 17.507 || ETA: 6 days, 19:01:42 || timer: 0.658

In the middle:
[ 16]  237920 || B: 1.659 | C: 2.578 | M: 2.639 | S: 0.531 | T: 7.406 || ETA: 4 days, 11:02:14 || timer: 0.685
[ 16]  237930 || B: 1.669 | C: 2.573 | M: 2.658 | S: 0.524 | T: 7.424 || ETA: 4 days, 11:03:43 || timer: 0.729
[ 16]  237940 || B: 1.663 | C: 2.600 | M: 2.681 | S: 0.539 | T: 7.483 || ETA: 4 days, 11:02:38 || timer: 0.722

Toward the end:
[ 41]  601220 || B: 1.330 | C: 1.869 | M: 2.277 | S: 0.362 | T: 5.838 || ETA: 2 days, 9:32:18 || timer: 0.629
[ 41]  601230 || B: 1.346 | C: 1.900 | M: 2.307 | S: 0.389 | T: 5.941 || ETA: 2 days, 9:32:34 || timer: 0.681
[ 41]  601240 || B: 1.362 | C: 1.919 | M: 2.314 | S: 0.396 | T: 5.991 || ETA: 2 days, 9:32:10 || timer: 0.666
```
(note that the ETAs are inaccurate in this log)

If you want to try to turn of data augmentation, you can add the following 4 settings to the `yolact_im700_config` in `data/config.py` (or pick and choose):
```python
# Randomize hue, vibrance, etc.
'augment_photometric_distort': False,
# Have a chance to scale down the image and pad (to emulate smaller detections)
'augment_expand': False,
# Potentialy sample a random crop from the image and put it in a random place
'augment_random_sample_crop': False,
# Mirror the image with a probability of 1/2
'augment_random_mirror': False,
```

Fair warning though, it will overfit on COCO if you turn off `augment_random_sample_crop`. 

Also if you're training with a batch size of 4, remember to freeze batch norm or bad things might happen. Add this to your config as well:
```Python
'freeze_bn': True,
```

**Comment by ridasalam** [comment]
Thanks.

So your total loss doesn't really go down much after? I wasn't sure if I had to change my hyper parameters.

**Comment by dbolya** [comment]
The total loss on COCO never goes below 5, so yeah. You shouldn't be looking at loss to see how the model is performing (since the losses are pretty abstract in what they mean). You should use validation mAP instead.

**Comment by ridasalam** [comment]
Ok thanks

**Comment by abhigoku10** [comment]
@ridasalam @dbolya  i trained my model with custom data with external augmentation and my loss is 6.0 range , should i increase the iterations for the loss to reduce up to 2-1

**Comment by dbolya** [comment]
Is that your total loss or individual losses?

**Comment by abhigoku10** [comment]
its the total loss

**Comment by dbolya** [comment]
That's going to be as low as it goes for total loss. On COCO I don't think it ever goes below like 5.7

**Comment by zhangbaoj** [comment]
How to draw the loss curve of YOLACT? It is  very important for me. Thank you very much.

**Comment by zhangbaoj** [comment]
How to draw the loss curve of YOLACT? It is  very important for me. Thank you very much.
And I don't know what is the mean of  B, C, M and T while training.

**Comment by dbolya** [comment]
@zhangbaoj see #78 for how to visualize the loss curve and what the individual losses mean.

**Comment by MiguelAngeloMartinsRibeiro** [comment]
> We do a lot of data augmentation which is likely why you can't get it to overfit. Here's some of the logs for im700:
> 
> ```
> Toward the start:
> [  0]     240 || B: 4.553 | C: 6.594 | M: 5.163 | S: 1.210 | T: 17.519 || ETA: 6 days, 20:00:21 || timer: 0.648
> [  0]     250 || B: 4.536 | C: 6.595 | M: 5.161 | S: 1.208 | T: 17.500 || ETA: 6 days, 19:30:03 || timer: 0.716
> [  0]     260 || B: 4.540 | C: 6.594 | M: 5.174 | S: 1.199 | T: 17.507 || ETA: 6 days, 19:01:42 || timer: 0.658
> 
> In the middle:
> [ 16]  237920 || B: 1.659 | C: 2.578 | M: 2.639 | S: 0.531 | T: 7.406 || ETA: 4 days, 11:02:14 || timer: 0.685
> [ 16]  237930 || B: 1.669 | C: 2.573 | M: 2.658 | S: 0.524 | T: 7.424 || ETA: 4 days, 11:03:43 || timer: 0.729
> [ 16]  237940 || B: 1.663 | C: 2.600 | M: 2.681 | S: 0.539 | T: 7.483 || ETA: 4 days, 11:02:38 || timer: 0.722
> 
> Toward the end:
> [ 41]  601220 || B: 1.330 | C: 1.869 | M: 2.277 | S: 0.362 | T: 5.838 || ETA: 2 days, 9:32:18 || timer: 0.629
> [ 41]  601230 || B: 1.346 | C: 1.900 | M: 2.307 | S: 0.389 | T: 5.941 || ETA: 2 days, 9:32:34 || timer: 0.681
> [ 41]  601240 || B: 1.362 | C: 1.919 | M: 2.314 | S: 0.396 | T: 5.991 || ETA: 2 days, 9:32:10 || timer: 0.666
> ```
> 
> If you want to try to turn off data augmentation, you can add the following 4 settings to the `yolact_im700_config` in `data/config.py` (or pick and choose):
> 
> ```python
> # Randomize hue, vibrance, etc.
> 'augment_photometric_distort': False,
> # Have a chance to scale down the image and pad (to emulate smaller detections)
> 'augment_expand': False,
> # Potentialy sample a random crop from the image and put it in a random place
> 'augment_random_sample_crop': False,
> # Mirror the image with a probability of 1/2
> 'augment_random_mirror': False,
> ```
> 
> Fair warning though, it will overfit on COCO if you turn off `augment_random_sample_crop`.
> 
> Also if you're training with a batch size of 4, remember to freeze batch norm or bad things might happen. Add this to your config as well:
> 
> ```python
> 'freeze_bn': True,
> ```

@dbolya do you advise to freeze batch norm if the batch size is smaller than 8? 

Does it affect the map of the network or can just screw training?

Thank you
